pkg/apitype/events.go

```diff
@@ -76,6 +76,24 @@ type SummaryEvent struct {
 	ResourceChanges map[string]int `json:"resourceChanges"`
 }
 
+// DiffKind describes the kind of a particular property diff.
+type DiffKind string
+
+const (
+	// DiffAdd indicates that the property was added.
+	DiffAdd DiffKind = "add"
+	// DiffAddReplace indicates that the property was added and requires that the resource be replaced.
+	DiffAddReplace DiffKind = "add-replace"
+	// DiffDelete indicates that the property was deleted.
+	DiffDelete DiffKind = "delete"
+	// DiffDeleteReplace indicates that the property was added and requires that the resource be replaced.
+	DiffDeleteReplace DiffKind = "delete-replace"
+	// DiffUpdate indicates that the property was updated.
+	DiffUpdate DiffKind = "update"
+	// DiffUpdateReplace indicates that the property was updated and requires that the resource be replaced.
+	DiffUpdateReplace DiffKind = "update-replace"
```

Comment:
My biggest question/concern around this PR is the complexity of this new contract between providers and the engine.  It feels like correctly computing these is going to be "hard" for a new provider.  

A few specific questions:
1. Have we verified that we can do this completely correctly for at least Terraform-based providers and Kubernetes providers?
2. Do we have any example of doing this for a non-schematized provider?  That is - for someone brining up a raw provider from scratch backed by e.g. a simple REST API - in practice how much net complexity does this add?
3. Should we be thinking about a schema model for native Pulumi providers that (among other things) gives this "for free"?  Something that allows a higher-level definition of new Pulumi providers (at least in Go)?

cc @jen20 @stack72 @lblackstone as well as they have experience on the other side of this contract.

Reply:
> My biggest question/concern around this PR is the complexity of this new contract between providers and the engine. It feels like correctly computing these is going to be "hard" for a new provider.

For what it's worth, it is not necessary for the provider to compute these if it can guarantee that the only diffs it sees are detectable by a simple diff of the old inputs against the new inputs (or the new inputs against the old state). If you can think of a better way to do this for providers that don't have this characteristic (e.g. TF providers), I am open to it.

> Have we verified that we can do this completely correctly for at least Terraform-based providers and Kubernetes providers?

For TF providers, yes. We should be able to do this for k8s as well by deriving the diff from a merge patch.

> Do we have any example of doing this for a non-schematized provider? That is - for someone brining up a raw provider from scratch backed by e.g. a simple REST API - in practice how much net complexity does this add?
> ...
> Should we be thinking about a schema model for native Pulumi providers that (among other things) gives this "for free"? Something that allows a higher-level definition of new Pulumi providers (at least in Go)?

I believe that these two things are so closely related that they really form one point. I do think that we should consider investing in some sort of schema model for native Pulumi providers, yes. https://github.com/pulumi/pulumi-openfaas/blob/master/pkg/provider/provider.go#L109-L119 is an example of one experimental approach using Go struct tags.

Reply:
> For TF providers, yes. We should be able to do this for k8s as well by deriving the diff from a merge patch.

I have now verified that we can do this in a rather straightforward fashion in the k8s provider. It is a little bit more complicated to include the `*Replace` annotations.

Reply:
I have now added support for the `*Replace` annotations in the k8s provider.